paddle/fluid/operators/elementwise/mkldnn/elementwise_mul_mkldnn_op.cc

```diff
@@ -130,6 +130,13 @@ class ElementwiseMulMKLDNNKernel : public framework::OpKernel<T> {
 
       z->set_layout(DataLayout::kMKLDNN);
       z->set_format(x->format());
+
+      // convert to nchw format to align with native version
+      using platform::MKLDNNDeviceContext;
+      auto& dev_ctx = ctx.template device_context<MKLDNNDeviceContext>();
+      const auto& mkldnn_engine = dev_ctx.GetEngine();
+      ReorderInput<T>(const_cast<Tensor*>(z), ctx.GetPlace(), mkldnn_engine,
+                      z->dims().size() == 4);
```

Comment:
I understand element-wise should not be aware of format, which means input format is equal to output format in op semantics. If adding the above code to fix UT failure but break the op semantics, we may need to rethink carefully.

Reply:
This is to align with native version, avx512 branch code is only case for.
```
    if (is_x_format_correct && is_y_format_correct && are_dims_divisable &&
        is_avx512_enabled)
```
which output format is different with other native version, so this patch is used to alignment for both output format and test case.

Reply:
currently, for native version, whatever x, y format is, the output is nchw or nc, while avx512 new added branch, the output is nchw16c, which made format difference on SKUs with different avx support

Reply:
My point is to follow the op semantics, namely element-wise op should not change the format (output format = input format), no matter how kernel implements and how UT works. Otherwise, it may introduce unnecessary reorders at topology level. E.g., conv1 (nchw16c) -> element-wise mul1; conv2 (nchw16c) -> element-wise mul1; element-wise mul1 -> conv3; for this pattern, it is expected to execute with nchw16c in the entire flow (no reorder from nchw to nchw16c). Add @tensor-tang for comments.

Reply:
> currently, for native version, whatever x, y format is, the output is nchw or nc, while avx512 new added branch, the output is nchw16c, which made format difference on SKUs with different avx support

From your description, it looks more like a test case issue, not kernel issue. Better to fix in the test code, instead of kernel.

Reply:
I agree, we shall avoid unnecessary reordering, e.g. output format follows x format.
I think that's the way for further code and test case optimization.

Reply:
> > currently, for native version, whatever x, y format is, the output is nchw or nc, while avx512 new added branch, the output is nchw16c, which made format difference on SKUs with different avx support
> 
> From your description, it looks more like a test case issue, not kernel issue. Better to fix in the test code, instead of kernel.

My opinion is test case should be transparent for low level SKU difference, and output format should be unified, that more makes sense.

Reply:
We are planning to prepare assembly version of this operator which supports AVX2 and AVX-512 VNNI. Maybe until then we could consider detecting CPU features in this UT with Python API (e.g. cpufeature module)

Reply:
Want to understand the rule of input/output format mapping, e.g. when x_data_format is nchw/nchw16c/xx, y is nchw/nchw16c/xx, what's the expected output format ? or it may be dynamic due to CPU SKU ? 

The patch here is just a quick solution to align avx512 version with naive version to unblock CI, may not be a long-term solution for giving consideration to both performance and interface definition.

Reply:
https://pypi.org/project/py-cpuinfo/
python for cpu info, if needed for detecting cpu features

Reply:
Yes, output format shall be equivalent to the x_data_format.

Reply:
Then, for naive version (working on non-avx512 machine), need follow same rule? or output the most performance-optimized format for specific CPU?

I was trying to convert format in naive version code, but found it is hard to do since it will break some existing test cases in mkldnn mul op.  Maybe need further adjustment on test cases?

Reply:
It is because the y_data is broadcasted to x_data during multiplications, and x_data defines the main path of the data flow. Usually elementwise_mul is used for scaling of the output parameters of some certain layer in a given topology. Since it is quite often the case that topologies such as ResNeXt-50 consist of repeating blocks of layers, it is most often the case that the input format to such block will be equivalent to the output format it yields (excluding the last block which is linked to the topology's output, e.g. the `fetch` op). Hence, keeping the same data format across almost the whole data path allows avoiding the reorders.

Reply:
I actually totally agree with you on the initiative. I think currently the only problem we need to solve is just to avoid the mismatch between different CPU SKU.
My confusion was, given the x input format is nchw16c on non-avx512 machine, what's the expect output format? your answer by my understanding is nchw, right? while I am thinking if we can unify and simply the rule between SKUs, e.g. output follows x input, it means naive version also outputs nchw16c by doing convert internally.

Reply:
nchw16c format shall be chosen by mkl-dnn engine only on avx512 compliant machines (so that they can use avx512 instructions to process binary operations on pairs of 16 floats at once), so at least in theory it should never occur that we encounter nchw16c on non-avx512 machine.

If magic happens, then you are right, the nchw16c would need to be converted to nchw so that the naive kernel can correctly process it. However, instead I would propose adding a PADDLE_ENFORCE that would verify that there is no nchw16c format on non-avx512 compliant machine, because if this occurs, something is clearly wrong.

Reply:
Good, this approach makes more sense. Then we may also need to update test case since current case design doesn't consider CPU flag difference, and better to disable this case on non-avx512 machine.

Reply:
Good idea, I totally agree. For now, in case of non-avx512 machine only the fallback test case should be executed (at least for until we introduce an avx2 kernel).